dallinger/recruiters.py

```diff
@@ -516,6 +534,99 @@ def _get_bot_factory(self):
         return Bot
 
 
+class MultiRecruiter(Recruiter):
+
+    nickname = 'multi'
+
+    # recruiter spec e.g. recruiters = bots: 5, mturk: 1
+    SPEC_RE = re.compile(r'(\w+):\s*(\d+)')
+
+    def __init__(self):
+        self.spec = self.parse_spec()
+
+    def parse_spec(self):
+        """Parse the specification of how to recruit participants.
+
+        Example: recruiters = bots: 5, mturk: 1
+        """
+        recruiters = []
+        spec = get_config().get('recruiters')
+        for match in self.SPEC_RE.finditer(spec):
+            name = match.group(1)
+            count = int(match.group(2))
+            recruiters.append((name, count))
+        return recruiters
+
+    def pick_recruiter(self):
+        """Pick the next recruiter to use.
+
+        We use the `Recruitment` table in the db to keep track of
+        how many recruitments have been requested using each recruiter.
+        We'll use the first one from the specification that
+        hasn't already reached its quota.
+        """
+        counts = dict(
+            session.query(
+                Recruitment.recruiter_id,
+                func.count(Recruitment.id)
+            ).group_by(Recruitment.recruiter_id).all()
+        )
+
+        for recruiter_id, target_count in self.spec:
+            count = counts.get(recruiter_id, 0)
+            if count >= target_count:
+                # This recruiter quota was reached;
+                # move on to the next one.
+                counts[recruiter_id] = count - target_count
```

Comment:
I couldn't figure out what this assignment did, and removing it didn't cause a test failure. Vestigial?

Reply:
This is to make sure that our bookkeeping is correct if the same recruiter is listed more than once in the spec, so that we skip over it as many times as it's already been used. For example you could have a spec like `bots: 1, cli: 1, bots: 1, cli: 1`